File: backend/src/capture.rs
use pcap::Capture;
use etherparse::{SlicedPacket, InternetSlice, TransportSlice, LinkSlice};
use crate::state::{SharedState, TrafficLog};
use crate::config::Config;
use crate::firewall::FirewallManager;
use crate::detection::{analyze_packet, PacketInfo};
use std::time::{SystemTime, UNIX_EPOCH};

pub fn start_capture(state: SharedState, config: Config, firewall: std::sync::Arc<FirewallManager>, handle: tokio::runtime::Handle) {
    let iface_name = config.interface.clone();
    
    std::thread::spawn(move || {
        println!("Starting capture on interface: {}", iface_name);
        let mut cap = match Capture::from_device(iface_name.as_str())
            .unwrap()
            .promisc(true)
            .snaplen(65535)
            .immediate_mode(true)
            .timeout(100)
            .open() {
                Ok(c) => c,
                Err(e) => {
                    eprintln!("Failed to open capture device {}: {:?}", iface_name, e);
                    return;
                }
            };

        let link_type = cap.get_datalink();
        println!("Datalink type: {:?}", link_type);

        loop {
            match cap.next_packet() {
                Ok(packet) => {
                    let ts = SystemTime::now()
                        .duration_since(UNIX_EPOCH)
                        .unwrap()
                        .as_secs_f64();

                    let mut pkt_info = PacketInfo {
                        src_ip: None,
                        dst_ip: None,
                        src_mac: None,
                        dst_mac: None,
                        protocol: "UNKNOWN".to_string(),
                        src_port: None,
                        dst_port: None,
                        is_tcp_syn: false,
                        is_tcp_ack: false,
                        is_tcp_rst: false,
                        payload: Vec::new(),
                    };

                    // Parse based on datalink type
                    let parsed_packet = if link_type == pcap::Linktype::LINUX_SLL && packet.data.len() >= 16 {
                        SlicedPacket::from_ip(&packet.data[16..]).ok()
                    } else if link_type == pcap::Linktype::NULL && packet.data.len() >= 4 {
                        SlicedPacket::from_ip(&packet.data[4..]).ok()
                    } else {
                        SlicedPacket::from_ethernet(packet.data).ok()
                    };

                    if let Some(parsed) = parsed_packet {
                        // Extract MAC addresses (only available for Ethernet link type)
                        if let Some(LinkSlice::Ethernet2(eth)) = parsed.link {
                            pkt_info.src_mac = Some(format!("{:02x}:{:02x}:{:02x}:{:02x}:{:02x}:{:02x}",
                                eth.source()[0], eth.source()[1], eth.source()[2],
                                eth.source()[3], eth.source()[4], eth.source()[5]));
                            pkt_info.dst_mac = Some(format!("{:02x}:{:02x}:{:02x}:{:02x}:{:02x}:{:02x}",
                                eth.destination()[0], eth.destination()[1], eth.destination()[2],
                                eth.destination()[3], eth.destination()[4], eth.destination()[5]));

                            if eth.ether_type() == 0x0806 {
                                pkt_info.protocol = "ARP".to_string();
                            }
                        }

                        // Extract IP addresses
                        if let Some(ip) = parsed.ip {
                            match ip {
                                InternetSlice::Ipv4(ipv4, _) => {
                                    pkt_info.src_ip = Some(ipv4.source_addr().to_string());
                                    pkt_info.dst_ip = Some(ipv4.destination_addr().to_string());
                                },
                                InternetSlice::Ipv6(ipv6, _) => {
                                    pkt_info.src_ip = Some(ipv6.source_addr().to_string());
                                    pkt_info.dst_ip = Some(ipv6.destination_addr().to_string());
                                }
                            }
                        }

                        // Extract transport layer info
                        if let Some(transport) = parsed.transport {
                            match transport {
                                TransportSlice::Tcp(tcp) => {
                                    pkt_info.protocol = "TCP".to_string();
                                    pkt_info.src_port = Some(tcp.source_port());
                                    pkt_info.dst_port = Some(tcp.destination_port());
                                    pkt_info.is_tcp_syn = tcp.syn();
                                    pkt_info.is_tcp_ack = tcp.ack();
                                    pkt_info.is_tcp_rst = tcp.rst();
                                },
                                TransportSlice::Udp(udp) => {
                                    pkt_info.protocol = "UDP".to_string();
                                    pkt_info.src_port = Some(udp.source_port());
                                    pkt_info.dst_port = Some(udp.destination_port());

                                    // Traffic Log Generation for UDP (DNS)
                                    if udp.destination_port() == 53 || udp.source_port() == 53 {
                                        extract_dns_log(&state, ts, &pkt_info.src_ip, parsed.payload, &handle);
                                    }
                                },
                                TransportSlice::Icmpv4(_) | TransportSlice::Icmpv6(_) => {
                                    pkt_info.protocol = "ICMP".to_string();
                                },
                                _ => {}
                            }
                        }

                        // Traffic log for TCP (HTTP/TLS)
                        if pkt_info.protocol == "TCP" && !parsed.payload.is_empty() {
                            extract_tcp_log(&state, ts, &pkt_info.src_ip, parsed.payload, &handle);
                        }
                    }

                    // Send packet to detection engine (even if parsing partially failed, src_ip might still be None and detection will skip it)
                    let state_clone = state.clone();
                    let config_clone = config.clone();
                    let firewall_clone = firewall.clone();
                    handle.spawn(async move {
                        analyze_packet(state_clone, &config_clone, &firewall_clone, pkt_info).await;
                    });
                },
                Err(pcap::Error::TimeoutExpired) => continue,
                Err(e) => {
                    eprintln!("Capture error: {:?}", e);
                    break;
                }
            }
        }
    });
}

fn extract_dns_log(state: &SharedState, ts: f64, src_ip: &Option<String>, payload: &[u8], handle: &tokio::runtime::Handle) {
    if payload.len() < 12 { return; }
    let qdcount = u16::from_be_bytes([payload[4], payload[5]]);
    if qdcount > 0 {
        let mut offset = 12;
        let mut domain = String::new();
        while offset < payload.len() {
            let len = payload[offset] as usize;
            if len == 0 { break; }
            if len > 63 { break; }
            offset += 1;
            if offset + len <= payload.len() {
                if !domain.is_empty() { domain.push('.'); }
                if let Ok(label) = std::str::from_utf8(&payload[offset..offset+len]) {
                    domain.push_str(label);
                }
            }
            offset += len;
        }
        if !domain.is_empty() {
            append_traffic_log(state, ts, src_ip, domain, "DNS", handle);
        }
    }
}

fn extract_tcp_log(state: &SharedState, ts: f64, src_ip: &Option<String>, payload: &[u8], handle: &tokio::runtime::Handle) {
    if let Ok(s) = std::str::from_utf8(payload) {
        if s.starts_with("GET ") || s.starts_with("POST ") || s.starts_with("PUT ") {
            for line in s.lines() {
                if line.to_lowercase().starts_with("host:") {
                    let domain = line[5..].trim().to_string();
                    append_traffic_log(state, ts, src_ip, domain, "HTTP", handle);
                    return;
                }
            }
        }
    }

    // Very basic TLS ClientHello SNI extraction
    if payload.len() > 43 && payload[0] == 0x16 && payload[1] == 0x03 && payload[5] == 0x01 {
        append_traffic_log(state, ts, src_ip, "TLS Session".to_string(), "HTTPS", handle);
    }
}

fn append_traffic_log(state: &SharedState, ts: f64, src_ip: &Option<String>, domain: String, proto: &str, handle: &tokio::runtime::Handle) {
    let src = src_ip.clone().unwrap_or_default();
    let proto_str = proto.to_string();
    let state = state.clone();

    handle.spawn(async move {
        let mut st = state.write().await;
        st.traffic_log.push_back(TrafficLog {
            timestamp: ts,
            src_ip: src,
            domain,
            proto: proto_str,
        });
        if st.traffic_log.len() > 200 {
            st.traffic_log.pop_front();
        }
    });
}

File: backend/capture/capture.go
package capture

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	"idps-backend/config"
	"idps-backend/detection"
	"idps-backend/firewall"
	"idps-backend/state"

	"github.com/google/gopacket"
	"github.com/google/gopacket/layers"
	"github.com/google/gopacket/pcap"
)

func getTimestamp() float64 {
	return float64(time.Now().UnixNano()) / 1e9
}

func StartCapture(st *state.AppState, cfg *config.Config, fm *firewall.FirewallManager, engine *detection.Engine) (func(), error) {
	ifaceName := cfg.CaptureInterface
	fmt.Printf("Starting capture on interface: %s\n", ifaceName)

	// Use a 1-second timeout instead of BlockForever to allow clean shutdown checks
	handle, err := pcap.OpenLive(ifaceName, 65535, true, 1*time.Second)
	if err != nil {
		return nil, fmt.Errorf("failed to open capture device %s: %w", ifaceName, err)
	}

	linkType := handle.LinkType()
	fmt.Printf("Datalink type: %v\n", linkType)

	packetSource := gopacket.NewPacketSource(handle, handle.LinkType())

	workerCount := cfg.WorkerCount
	if workerCount <= 0 {
		workerCount = 4
	}

	channels := make([]chan detection.PacketInfo, workerCount)
	var wg sync.WaitGroup

	for i := 0; i < workerCount; i++ {
		channels[i] = make(chan detection.PacketInfo, 2000)
		wg.Add(1)
		go func(ch chan detection.PacketInfo) {
			defer wg.Done()
			for pkt := range ch {
				engine.ProcessPacket(pkt)
			}
		}(channels[i])
	}

	ctx, cancel := context.WithCancel(context.Background())

	go func() {
		defer func() {
			fmt.Printf("Capture stopped on interface: %s\n", ifaceName)
			for i := 0; i < workerCount; i++ {
				close(channels[i])
			}
		}()

		for {
			select {
			case <-ctx.Done():
				return
			default:
			}

			packet, err := packetSource.NextPacket()
			if err != nil {
				// pcap.NextErrorTimeoutExpired is expected every 1s if idle
				if strings.Contains(err.Error(), "Timeout Expired") || err.Error() == "Timeout Expired" {
					continue
				}
				// If handle is closed, err will be EOF or a closed error
				return
			}

			if packet == nil {
				continue
			}

			ts := getTimestamp()

			pktInfo := detection.PacketInfo{
				Protocol: "UNKNOWN",
			}

			// Ethernet
			ethLayer := packet.Layer(layers.LayerTypeEthernet)
			if ethLayer != nil {
				eth, _ := ethLayer.(*layers.Ethernet)
				pktInfo.SrcMAC = eth.SrcMAC.String()
				pktInfo.DstMAC = eth.DstMAC.String()
				if eth.EthernetType == layers.EthernetTypeARP {
					pktInfo.Protocol = "ARP"
					arpLayer := packet.Layer(layers.LayerTypeARP)
					if arpLayer != nil {
						arp, _ := arpLayer.(*layers.ARP)
						pktInfo.ARPOperation = arp.Operation
					}
				}
			} else {
				// Try Linux SLL
				sllLayer := packet.Layer(layers.LayerTypeLinuxSLL)
				if sllLayer != nil {
					sll, _ := sllLayer.(*layers.LinuxSLL)
					if len(sll.Addr) >= 6 {
						pktInfo.SrcMAC = fmt.Sprintf("%02x:%02x:%02x:%02x:%02x:%02x", sll.Addr[0], sll.Addr[1], sll.Addr[2], sll.Addr[3], sll.Addr[4], sll.Addr[5])
					}
				}
			}

			// IP
			ip4Layer := packet.Layer(layers.LayerTypeIPv4)
			if ip4Layer != nil {
				ip4, _ := ip4Layer.(*layers.IPv4)
				pktInfo.SrcIP = ip4.SrcIP.String()
				pktInfo.DstIP = ip4.DstIP.String()
			} else {
				ip6Layer := packet.Layer(layers.LayerTypeIPv6)
				if ip6Layer != nil {
					ip6, _ := ip6Layer.(*layers.IPv6)
					pktInfo.SrcIP = ip6.SrcIP.String()
					pktInfo.DstIP = ip6.DstIP.String()
				}
			}

			// Transport
			tcpLayer := packet.Layer(layers.LayerTypeTCP)
			if tcpLayer != nil {
				tcp, _ := tcpLayer.(*layers.TCP)
				pktInfo.Protocol = "TCP"
				pktInfo.SrcPort = uint16(tcp.SrcPort)
				pktInfo.DstPort = uint16(tcp.DstPort)
				pktInfo.IsTCPSYN = tcp.SYN
				pktInfo.IsTCPACK = tcp.ACK
				pktInfo.IsTCPRST = tcp.RST
				pktInfo.Seq = tcp.Seq
				pktInfo.Payload = tcp.Payload

				if len(tcp.Payload) > 0 {
					extractTCPLog(st, ts, pktInfo.SrcIP, tcp.Payload)
				}
			} else {
				udpLayer := packet.Layer(layers.LayerTypeUDP)
				if udpLayer != nil {
					udp, _ := udpLayer.(*layers.UDP)
					pktInfo.Protocol = "UDP"
					pktInfo.SrcPort = uint16(udp.SrcPort)
					pktInfo.DstPort = uint16(udp.DstPort)
					pktInfo.Payload = udp.Payload

					if udp.DstPort == 53 || udp.SrcPort == 53 {
						extractDNSLog(st, ts, pktInfo.SrcIP, udp.Payload)
					}
				} else {
					icmp4Layer := packet.Layer(layers.LayerTypeICMPv4)
					if icmp4Layer != nil {
						pktInfo.Protocol = "ICMP"
						icmp, _ := icmp4Layer.(*layers.ICMPv4)
						pktInfo.Payload = icmp.Payload
					} else {
						icmp6Layer := packet.Layer(layers.LayerTypeICMPv6)
						if icmp6Layer != nil {
							pktInfo.Protocol = "ICMP"
							icmp, _ := icmp6Layer.(*layers.ICMPv6)
							pktInfo.Payload = icmp.Payload
						}
					}
				}
			}

			if pktInfo.SrcIP != "" {
				var hash uint32
				for i := 0; i < len(pktInfo.SrcIP); i++ {
					hash = hash*31 + uint32(pktInfo.SrcIP[i])
				}
				idx := hash % uint32(workerCount)
				
				select {
				case channels[idx] <- pktInfo:
				default:
					st.Mu.Lock()
					st.DroppedPacketCount++
					st.Mu.Unlock()
				}
			}
		}
	}()

	var stopOnce sync.Once
	stop := func() {
		stopOnce.Do(func() {
			cancel() // Signal the capture loop to stop
			handle.Close()
			wg.Wait()
		})
	}

	return stop, nil
}

func extractDNSLog(st *state.AppState, ts float64, srcIP string, payload []byte) {
	if len(payload) < 12 {
		return
	}
	// Parse basic DNS header to see if QDCOUNT > 0
	qdcount := uint16(payload[4])<<8 | uint16(payload[5])
	if qdcount > 0 {
		offset := 12
		var domainParts []string
		for offset < len(payload) {
			length := int(payload[offset])
			if length == 0 || length > 63 {
				break
			}
			offset++
			if offset+length <= len(payload) {
				domainParts = append(domainParts, string(payload[offset:offset+length]))
			}
			offset += length
		}
		domain := strings.Join(domainParts, ".")
		if domain != "" {
			appendTrafficLog(st, ts, srcIP, domain, "DNS")
		}
	}
}

func extractTCPLog(st *state.AppState, ts float64, srcIP string, payload []byte) {
	s := string(payload)
	if strings.HasPrefix(s, "GET ") || strings.HasPrefix(s, "POST ") || strings.HasPrefix(s, "PUT ") {
		lines := strings.Split(s, "\r\n")
		for _, line := range lines {
			if strings.HasPrefix(strings.ToLower(line), "host:") {
				domain := strings.TrimSpace(line[5:])
				appendTrafficLog(st, ts, srcIP, domain, "HTTP")
				return
			}
		}
	}

	// Basic TLS ClientHello SNI extraction approximation
	if len(payload) > 43 && payload[0] == 0x16 && payload[1] == 0x03 && payload[5] == 0x01 {
		appendTrafficLog(st, ts, srcIP, "TLS Session", "HTTPS")
	}
}

func appendTrafficLog(st *state.AppState, ts float64, srcIP, domain, proto string) {
	if srcIP == "" {
		return
	}

	st.Mu.Lock()
	defer st.Mu.Unlock()
	st.AddTrafficLog(state.TrafficLog{
		Timestamp: ts,
		SrcIP:     srcIP,
		Domain:    domain,
		Proto:     proto,
	})
}
